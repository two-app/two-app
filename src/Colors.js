// @flow

const Colors = {
    SALMON: "#EE7674",
    DARK_SALMON: "#ed615f",
    LIGHT: "rgba(0, 0, 0, 0.1)",
    FADED: "rgba(0, 0, 0, 0.2)",
    DARK: "rgba(0, 0, 0, 0.8)",
    VALID_GREEN: "#85CC73"
};

export default Colors;