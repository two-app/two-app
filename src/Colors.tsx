const Colors = {
  SALMON: '#eb5754',
  DARK_SALMON: '#ed615f',
  LIGHT: 'rgba(0, 0, 0, 0.1)',
  FADED: 'rgba(0, 0, 0, 0.2)',
  REGULAR: 'rgb(100, 100, 100)',
  DARK: 'rgb(37, 37, 37)',
  VERY_DARK: 'rgb(21, 21, 21)',
  VALID_GREEN: '#3AB795',
  VALID_GREEN_DARK: '#36a88a',
};

export default Colors;
