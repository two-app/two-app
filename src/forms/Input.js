// @flow
import React, {useState} from "react";
import PropTypes from "prop-types";
import {StyleSheet, TextInput} from "react-native";
import Colors from "../Colors";

/**
 * @callback isValidFn
 * @param {string} text in the input.
 * @returns {boolean} true if the input is valid.
 */

/**
 * Emit is called when the user deselects the input.
 * @callback onEmitFn
 * @param {string} the last valid value in the input.
 */

/**
 * @param {Object} attributes to apply on the input, e.g {placeholder: "email"}
 * @param {isValidFn} isValid callback to check if the input text is valid.
 * @param {onEmitFn} onEmit callback when the user deselects the input.
 * @returns {*} the Input component.
 */
const Input = ({attributes, isValid, onEmit}) => {
    const [value, setValue] = useState("");
    const [valid, setValid] = useState(true);

    return <TextInput
        {...attributes}
        onChangeText={setValue}
        onBlur={() => {
            const isValueValid = isValid(value);
            setValid(isValueValid);
            if (isValueValid) onEmit(value);
        }}
        style={[styles.input]}
    />;
};

Input.propTypes = {
    attributes: PropTypes.object,
    isValid: PropTypes.func,
    onEmit: PropTypes.func
};

Input.defaultProps = {
    attributes: {},
    isValid: () => true,
    onEmit: () => {
    }
};

const styles = StyleSheet.create({
    input: {
        width: '100%',
        borderBottomColor: Colors.FADED,
        borderBottomWidth: 1,
        paddingHorizontal: 0,
        paddingTop: 10,
        paddingBottom: 10,
        fontSize: 15
    }
});

export default Input;