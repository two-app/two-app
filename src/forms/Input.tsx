import {forwardRef, ReactElement, useState} from 'react';
import {
  StyleSheet,
  TextInput,
  TextInputProps,
  View,
  ViewStyle,
  Text,
} from 'react-native';
import {TextInputMask, TextInputMaskProps} from 'react-native-masked-text';
import {Icon} from 'react-native-vector-icons/Icon';

import Colors from '../Colors';

type ValidatedString = [string, boolean];

type InputProps = TextInputProps & {
  initialValue?: string;
  onEmit: (e: ValidatedString) => void;
  isValid?: (value: string) => boolean;
  icon?: {
    name: string;
    provider: typeof Icon;
  };
  mask?: TextInputMaskProps;
  containerStyle?: ViewStyle;
};

export const Input = forwardRef((props: InputProps, ref) => {
  const [value, setValue] = useState<string>(props.initialValue ?? '');
  const [valid, setValid] = useState<boolean>(true);
  const [focused, setFocused] = useState<boolean>(false);
  let focusedStyle;

  if (focused) {
    focusedStyle = styles.focused;
  } else if (value === '') {
    focusedStyle = styles.unfocused;
  } else {
    focusedStyle = valid ? styles.valid : styles.invalid;
  }

  const emit = () => {
    setFocused(false);
    const valid = props.isValid?.(value) ?? true;
    setValid(valid);
    props.onEmit([value, valid]);
  };

  const commonProps: TextInputProps = {
    value,
    onChangeText: setValue,
    onSubmitEditing: emit,
    onBlur: emit,
    onFocus: () => setFocused(true),
    style: [styles.input, focusedStyle],
    accessibilityValue: {text: valid ? 'Valid entry' : 'Invalid entry'},
    placeholderTextColor: 'rgba(0, 0, 0, 0.5)',
  };

  let InputComponent: ReactElement;

  if (props.mask) {
    InputComponent = (
      // @ts-ignore
      <TextInputMask {...commonProps} {...props.mask} {...props} ref={ref} />
    );
  } else {
    // @ts-ignore
    InputComponent = <TextInput {...commonProps} {...props} ref={ref} />;
  }

  return (
    <View style={[styles.container, props.containerStyle]}>
      {InputComponent}
      <View style={{...focusedStyle, ...styles.iconContainer}}>
        {props.icon != null && (
          <props.icon.provider
            name={props.icon.name}
            size={15}
            color={focusedStyle.color}
          />
        )}
      </View>
    </View>
  );
});

type FakeInputProps = {
  placeholder: string;
  value?: string;
  valid: boolean;
  icon?: {
    name: string;
    provider: typeof Icon;
  };
  containerStyle?: ViewStyle;
};

export const FakeInput = (props: FakeInputProps) => {
  const {value, valid, icon, containerStyle} = props;
  let focusStyle;

  if (value == null || value == '') {
    focusStyle = styles.unfocused;
  } else if (valid) {
    focusStyle = styles.valid;
  } else {
    focusStyle = styles.invalid;
  }

  let DataComponent: ReactElement;

  if (value == null || value == '') {
    // Create a placeholder
    DataComponent = (
      <Text style={{color: PLACEHOLDER_COLOR}}>{props.placeholder}</Text>
    );
  } else {
    DataComponent = <Text style={focusStyle}>{value}</Text>;
  }

  return (
    <View style={[styles.container, props.containerStyle]}>
      <View style={[styles.input, focusStyle]}>{DataComponent}</View>
      <View style={[focusStyle, styles.iconContainer]}>
        {props.icon != null && (
          <props.icon.provider
            name={props.icon.name}
            size={15}
            color={focusStyle.color}
          />
        )}
      </View>
    </View>
  );
};

const HEIGHT = 50;
const BORDER_WIDTH = 1;
const BORDER_RADIUS = 5;
const PLACEHOLDER_COLOR = 'rgba(0, 0, 0.0980392, 0.5)';
const FONT_SIZE = 14;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    height: HEIGHT,
    backgroundColor: 'white',
  },
  input: {
    flexGrow: 1,
    padding: 15,
    fontSize: FONT_SIZE,
    borderTopLeftRadius: BORDER_RADIUS,
    borderBottomLeftRadius: BORDER_RADIUS,
    borderTopWidth: BORDER_WIDTH,
    borderLeftWidth: BORDER_WIDTH,
    borderBottomWidth: BORDER_WIDTH,
  },
  iconContainer: {
    width: HEIGHT,
    height: HEIGHT,
    justifyContent: 'center',
    alignItems: 'center',
    borderTopRightRadius: BORDER_RADIUS,
    borderBottomRightRadius: BORDER_RADIUS,
    borderTopWidth: BORDER_WIDTH,
    borderRightWidth: BORDER_WIDTH,
    borderBottomWidth: BORDER_WIDTH,
  },
  unfocused: {
    borderColor: 'rgba(0, 0, 0, 0.2)',
    color: Colors.REGULAR,
  },
  focused: {
    borderColor: Colors.DARK,
    color: Colors.DARK,
  },
  invalid: {
    borderColor: Colors.SALMON,
    color: Colors.SALMON,
  },
  valid: {
    borderColor: Colors.VALID_GREEN,
    color: Colors.VALID_GREEN,
  },
});
