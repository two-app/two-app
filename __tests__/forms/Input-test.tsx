import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {shallow, ShallowWrapper} from 'enzyme';
import Input from '../../src/forms/Input';
import Colors from '../../src/Colors';


describe('Input', () => {
    let tb: InputTestBed;
    beforeEach(() => {
        tb = new InputTestBed();
    });

    test('should maintain snapshot', () => expect(renderer.create(<Input/>)).toMatchSnapshot());

    test('starts with an empty value', () => expect(tb.input.prop('value')).toEqual(''));

    test('applies supplied attributes', () => expect(tb.input.prop('placeholder')).toEqual(tb.attributes.placeholder));

    test('maintains the input value', () => {
        expect(tb.getInputValue()).toEqual('');
        tb.setInputValue('New Value');
        expect(tb.getInputValue()).toEqual('New Value');
    });
});

describe('On Change', () => {
    test('emits to onChangeFn', () => {
        const tb = new InputTestBed();

        tb.setInputValue('testVal');

        expect(tb.onChangeFn).toHaveBeenCalledWith('testVal');
    });
});

describe('On Blur', () => {
    let tb: InputTestBed;
    beforeEach(() => tb = new InputTestBed());

    test('checks if input is valid', () => {
        tb.blurInput('Test Text');
        expect(tb.isValidFn).toHaveBeenCalledWith('Test Text');
    });

    test('gets a dark border when focused', () => {
        tb.input.prop<() => void>('onFocus')();
        tb.refresh();
        expect(tb.input.prop<any>('style')[1]).toHaveProperty('borderBottomColor', Colors.DARK);
    });

    test('gets a salmon border when invalid', () => {
        tb.isValidFn.mockReturnValue(false);
        tb.blurInput('');
        expect(tb.input.prop<any>('style')[2]).toHaveProperty('borderBottomColor', Colors.SALMON);
    });
});

describe('With label', () => {
    test('creates the label', () => expect(new InputTestBed('lbl').wrapper.exists('Label')).toBe(true));

    test('passes label text in', () => {
        expect(new InputTestBed('lbl').wrapper.find('Label').prop('text')).toEqual('lbl');
    });

    test('does not create a label if none provided', () => {
        expect(new InputTestBed().wrapper.exists('Label')).toBe(false);
    });
});

class InputTestBed {

    attributes = {placeholder: 'Test Placeholder'};
    isValidFn = jest.fn();
    onChangeFn = jest.fn();
    wrapper: ShallowWrapper;
    input: ShallowWrapper;

    constructor(label?: string) {
        this.wrapper = shallow(
            <Input attributes={this.attributes} isValid={this.isValidFn} onChange={this.onChangeFn} label={label}/>
        );

        console.log(this.wrapper.debug())

        this.input = this.wrapper.find('TextInput');
    }

    refresh = () => {
        this.wrapper.update();
        this.input = this.wrapper.find('TextInput');
    };

    setInputValue = (value: string) => {
        this.input.prop<(v: string) => void>('onChangeText')(value);
        this.refresh();
    };

    blurInput = (value: string) => {
        this.setInputValue(value);
        this.input.prop<() => void>('onBlur')();
        this.refresh();
    };

    getInputValue = () => this.input.prop('value');
}