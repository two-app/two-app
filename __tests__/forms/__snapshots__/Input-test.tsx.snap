// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should maintain snapshot 1`] = `
<TextInput
  accessibilityValue={
    Object {
      "text": "Valid entry",
    }
  }
  onBlur={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  style={
    Array [
      Object {
        "borderBottomColor": "rgba(0, 0, 0, 0.2)",
        "borderBottomWidth": 1,
        "fontSize": 15,
        "paddingBottom": 10,
        "paddingHorizontal": 0,
        "paddingTop": 10,
        "width": "100%",
      },
      undefined,
      undefined,
    ]
  }
  value=""
/>
`;
